### PR TITLE
SLD import/export support for raster marker symbol layers

### PR DESCRIPTION
**Pull request by olivierdalang**
## Description

This adds support for raster marker symbol layers in SLD import/exports.

A few notes:
- For embedded files, it embed the file as a data uri in the SLD (for remote files, it keeps the URL as is)
- For exports, the mime type will be hardcoded to the generic `application/octet-stream`. I don't think QGIS knows the mime type, neither for remote files nor for embedded files. It seems to work with openlayers and `@nieuwlandgeo/sldreader`, I tested with gif and pngs. The the mime type element is mandatory in the spec, but doesn't really say much about how the mime type is used, esp. since it seems redundant for remote files (as the type should be sent throught the http header) as well as for embedded files (since the mime type is included in the data uri, which doesn't help in our case as we don't set it anyways).
- It doesn't cover the whole SLD spec regarding raster ExternalGraphic (which also has optional attributes for Opacity, Size, Rotation, AnchorPoint, Displacement). For now they are not exported, and ignored on import. I think these could be added later if needed, as they are not as simple as one could think (with all these data defined cases, the fact that size can be set at the symbol or at the symbol layer level, etc.), and just having the symbol kept is already a significant improvement.
- I haven't contributed in a long time, sorry in advance if I forgot some obvious stuff :-)

Questions:
- I guess this type of fix isn't elligible for backporting, correct ? (it could be presented as a fix for incomplete SLD import export)
- Regarding that mime type question, is the fact that it works with `@nieuwlandgeo/sldreader` good enough ? Or is there some reference implementation to check against ? (I guess geoserver ?). I've never used it, if someone who has it running could test it (with the SLD below) that would be very welcome !

Cheers !!

## Progress

- [x] base implementation
- [x] fix mime type in embedded resources (by resaving the image to a know format)
- [ ] ~wait (a little)&see what geoserver people say about the [data uris crashing in geoserver](https://osgeo-org.atlassian.net/browse/GEOS-11821)~ we don't use data uri anymore but standard SLD's InlineContent
- [ ] fix CI

SLD export before:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<StyledLayerDescriptor xmlns="http://www.opengis.net/sld" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" version="1.1.0" xmlns:ogc="http://www.opengis.net/ogc" xmlns:xlink="http://www.w3.org/1999/xlink" xsi:schemaLocation="http://www.opengis.net/sld http://schemas.opengis.net/sld/1.1.0/StyledLayerDescriptor.xsd" xmlns:se="http://www.opengis.net/se">
  <NamedLayer>
    <se:Name>TREES_LAYER</se:Name>
    <UserStyle>
      <se:Name>TREES_LAYER</se:Name>
      <se:FeatureTypeStyle>
        <se:Rule>
          <se:Name>Trees</se:Name>
          <se:Description>
            <se:Title>Trees</se:Title>
          </se:Description>
          <se:PointSymbolizer>
            <!--QgsMarkerSymbolLayer RasterMarker not implemented yet-->
          </se:PointSymbolizer>
        </se:Rule>
      </se:FeatureTypeStyle>
    </UserStyle>
  </NamedLayer>
</StyledLayerDescriptor>
```

SLD export after:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<StyledLayerDescriptor xmlns="http://www.opengis.net/sld" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" version="1.1.0" xmlns:ogc="http://www.opengis.net/ogc" xmlns:xlink="http://www.w3.org/1999/xlink" xsi:schemaLocation="http://www.opengis.net/sld http://schemas.opengis.net/sld/1.1.0/StyledLayerDescriptor.xsd" xmlns:se="http://www.opengis.net/se">
  <NamedLayer>
    <se:Name>TREES_LAYER</se:Name>
    <UserStyle>
      <se:Name>TREES_LAYER</se:Name>
      <se:FeatureTypeStyle>
        <se:Rule>
          <se:Name>Trees</se:Name>
          <se:Description>
            <se:Title>Trees</se:Title>
          </se:Description>
          <se:PointSymbolizer>
            <se:Graphic>
              <se:ExternalGraphic>
                <se:OnlineResource xlink:type="simple" xlink:href="data:application/octet-stream;base64,iVBORw0KGgoAAAANSUhEUgAAAAsAAAALCAYAAACprHcmAAAACXBIWXMAAA7EAAAOxAGVKw4bAAAA0klEQVQYlWNhIAGwkK244WooG8MjgSBGpv+CLMwM66pd5rzEqrhyXZwaPyPPJiFnDnVGFkaGz49/tbbsScmscZmzEkOxoBxXl4gmpzqMzyfHJvj/P8OE0FDLTatXH/+OopiDn9kA3Y18cmwSMs7K9gyrj+9AUfzvL8MrBgYGeWTF///8//v+5s+nGM54eenbfCE1DmNGRgYmmNjzs18XL+hffQVDMfs1+ZlXlt7/Ia7PFcDIxMj95cnvQycnP5sMsgBDcUNDwz8GBob5UEw4nAkBABe5Qtss89RNAAAAAElFTkSuQmCC"/>
                <se:Format>application/octet-stream</se:Format>
              </se:ExternalGraphic>
            </se:Graphic>
          </se:PointSymbolizer>
        </se:Rule>
      </se:FeatureTypeStyle>
    </UserStyle>
  </NamedLayer>
</StyledLayerDescriptor>
```

### CODE REVIEW

**Comment by nyalldawson** [comment]
This is a very nice improvement @olivierdalang ! 

I think we should test against geoserver just because it's probably the most common destination for exported slds. It's quite easy to test against -- there's a docker image you can use and it's quite strange to get a preview render of an sld via it.

**Comment by olivierdalang** [comment]
Ok I tested with geoserver, it's not looking good (at all).

**Remote URLs:**
- without the `<se:Format>` (which is an invalid sld) -> `NullPointerException`
- with `<se:Format></se:Format>` (I think valid) -> `NullPointerException`
- with `<se:Format>application/octet-stream</se:Format>` -> renders the default marker (same as with an empty `<se:PointSymbolizer>`)
- with `<se:Format>image/png</se:Format>` -> this works

I don't think including the mime type for remote URLs is justified, as it would require making web requests to retrieve it from upstream servers (since on qgsmarkersymbollayer we only have access to a QImage). Storing the original mime type in `QgsImageCache` may be too big of a change to work around geoserver quirks (also there's no guarantee that a specific URL always returns the same mime type, so it's really a weird design choice of SLD).

But I would say it's OK to keep application/octet-stream as at least there's no regression for geoserver.

**Remote URLs:**

Here it's worse independently from the mime type we get `RuntimeException: [...] unknown protocol: data`.

So in that case, it's worse than currently. I see no real work around here. As far as I understand, geoserver is at fault, as data uris are valid uris according to xlink. Here I'll open an issue on geoserver side to either support it (as it's really an useful feature) or at least not crash and just ignore the link.

So not really sure what we should do here. Until fixed by geoserver, that would be a regression. I guess no one currently exports embedded raster markers to geoserver since they'd just get the default style anyways, but still the case could happen for batch exports.

@nyalldawson Let me know what you think, I'm not really able to judge what's best here.

Otherwise will look into the CI failure, but I don't understand `PyQgsDocCoverage`'s failure `AssertionError: Error: Missing Q_DECL_DEPRECATED for void QgsRasterMarkerSymbolLayer::writeSldMarker`. Can you hint at what's expected ?

File: tests/src/python/test_qgssymbollayer_readsld.py [hunk 1]
```diff
@@ -616,6 +619,30 @@ def _check_layer(layer, yMargin=10, xMargin=15):
...
+        """Test reading raster markers"""
+        file_path = os.path.join(
+            TEST_DATA_DIR, "symbol_layer/QgsRasterMarkerSymbolLayer-remote.sld"
+        )
+        maplayer = QgsVectorLayer("Point?crs=epsg:3111&field=pk:int", "vl", "memory")
+        maplayer.loadSldStyle(file_path)
+        marker = maplayer.renderer().symbol().symbolLayers()[0]
+        self.assertEqual(marker.__class__, QgsRasterMarkerSymbolLayer)
+        # import pdb; pdb.set_trace()
+        # raise Exception(marker)
```
Comment (olivierdalang) [inline tests/src/python/test_qgssymbollayer_readsld.py:632]: cleanup

**Comment by nyalldawson** [comment]
@olivierdalang 

Hm, interesting. A potential solution could be to use the QImage from the symbol layer, and export that to a known format (eg PNG) and then embed THAT data, instead of the original image data. This way we'd always have a known fixed mime type, which should make geoserver happy. And it has the additional benefit of working for remote images too (although remote images would then become embedded images, but I think that's slightly preferred anyway). What do you think?

**Comment by olivierdalang** [comment]
No unfortunately, according to my testings, **embedded files don't work at all in geoserver**, even with correct mime types. So that wouldn't really solve anything for geoserver. It could still be worth it for other clients, so we could include an accurate mime type (both in the data uri itself and in SLD's Format element).

As for embedding remote images, I can see many scenarios where this could not be wanted (same symbols used in a lot of many symbologies, very large symbols, offline conversion, export then reimport idempotency, etc.). I find the conversion more true if we keep what's done in QGIS. I guess in most cases the arguments to decide between embedded vs remote symbols apply in the same way at the QGIS level already, so if one wants them embedded in the SLD, he may also want to embedded them in the QGIS project.

**Comment by olivierdalang** [comment]
I had another look, actually Qt has some helpers to guess mime types (hehe Qt really does everything). I pushed some new commits that include this, so that now mime types are correctly set for local and embedded files. For remote files however, Qt's helpers don't help, as they return the generic type, which break geoserver's rendering. However, geoserver is happy with wrong mimes, so let's just hardcode this to image/png.

I also tested for local files, and it works well. They can also be uploaded to geoserver, and I guess it's the most common use case.

So all in all, the only drawback I see of this MR is that embedded files crash geoserver's render while prior to this PR, they would render with the default renderer. IMO it's quite an edge case probably not used much in the wild, and could easily be fixed in geoserver (see [GEOS-11821](https://osgeo-org.atlassian.net/browse/GEOS-11821) ).

Worth noting that in my tests, some valid remote URLs with correct mime type rendered correctly with geoserver:2.23.0 but error with geoserver:2.27.0, so there maybe some more issues they have to sort out with SLD ?

**Comment by nyalldawson** [comment]
@olivierdalang 
looking at https://osgeo-org.atlassian.net/browse/GEOS-11821 I think there's an alternative (geoserver friendly) way we could handle embedded rasters. From https://github.com/qgis/QGIS/issues/37987#issuecomment-1746306826 embedded rasters could be formatted like:


````
  <se:ExternalGraphic>
      <se:InlineContent encoding="base64">iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAIAAACQkWg2AAAAK3RFWHRDcmVhdGlvbiBUaW1lAFd0IDE0IHdyeiAyMDEwIDEyOjA2OjAyICswMTAweoAlkgAAAAd0SU1FB9oJDgo6HdmGt90AAAAJcEhZcwAACxIAAAsSAdLdfvwAAAAEZ0FNQQAAsY8L/GEFAAABfklEQVR42mP8//8/AwYACn779o2bmxtTiokBG9iwboOfq9+1K9eI0vDp06cZk2e8ffp2zuQ5mPZj0bB963aGLwxTI6YdO3QMiyX/UcHz58/tLeyXZ674O+NfgVthWnTaz58/kRWg27B8yXKu31yeGp6MjIwpVimnjp86fvg4Tic9ffp0zfI1ZU7lfBx8QK6WpFasSdzMiTN//fqFRQPQusULFsvzyFsqWMIFw4zDHt5+uH/Xfiwarl27tnj+4nSzdDYWNrigtIB0iF5IZ1MnMOhQNACNnztjrq64roWiJQMqCDUO+/D6w/YN21E0XLp06fD+w3l2+cjGwy1Jt82YO3Pe+/fvoRqAfupq6bKVs7VQsMAa8REmEf8+/V04fSFUw/Hjxy+cvZBklgwMSqwagIGWYpm6cOEiYDCyAI2fPnG6mrD6n39/Lj25xIADAB325/vvVQtXsXz58uX7z+8vfr7I2ZrNgBew87HfvnebERg+X79+/ffvHwMRgJWVFQBa4Mt756r78AAAAABJRU5ErkJggg==</se:InlineContent>
      <se:Format>image/png</se:Format>
  </se:ExternalGraphic>
```

**Comment by olivierdalang** [comment]
Oh nice! unsure why I missed that. The only drawback is that it looks like @nieuwlandgeo/sldreader doesn't support it, but now that can be solved there I think.

Now one thing that could be improved is to factorize this with SVG, so that we can also embbed SVGs. I think it's doable, but requires a bit of refactoring, for instance `QgsSymbolLayerUtils::createOnlineResourceElement` would actually conditionaly create InlineContent (so worth a rename) and it should not call `svgSymbolPathToName` (so it can also be used for non svgs).

Is it OK to keep that as a future improvement ? (also minimizing risks in regard to FF exemption).

Also still having this failure in CI that I don't understand: ` AssertionError: Error: Missing Q_DECL_DEPRECATED for void QgsRasterMarkerSymbolLayer::writeSldMarker` (I reused the exact same definition as other subclasses... what's expected ?)

**Comment by nyalldawson** [comment]
@olivierdalang 

> Oh nice! unsure why I missed that. The only drawback is that it looks like neither @nieuwlandgeo/sldreader nor geostyler support it, but that can be solved there I think.

We could add a Qgis::SldExportOption to control this, and in future offer a user-facing choice as to whether to use InlineContent or embedded base64. (Potentially we could use the existing Qgis::SldExportVendorExtension::GeoServerVendorExtension option for this, but I think that's a little messy as the documentation for that enum all relates specifically to vendor extensions, not server compatibility quirks for standard SLD components...)

> Now one thing that could be improved is to factorize this with SVG, so that we can also embbed SVGs. I think it's doable, but requires a bit of refactoring, for instance QgsSymbolLayerUtils::createOnlineResourceElement would actually conditionaly create InlineContent (so worth a rename) and it should not call svgSymbolPathToName (so it can also be used for non svgs).

I'd ignore svgs initially -- for Geoserver at least embedded SVG is NOT supported (https://osgeo-org.atlassian.net/browse/GEOT-5206).  Getting this working for raster alone is a massive step forward and I think we should handle SVGs in a later follow up.

> Also still having this failure in CI that I don't understand:  AssertionError: Error: Missing Q_DECL_DEPRECATED for void QgsRasterMarkerSymbolLayer::writeSldMarker (I reused the exact same definition as other subclasses... what's expected ?)

Try rebasing your branch on master -- you'll need to update following https://github.com/qgis/QGIS/pull/61683

**Comment by olivierdalang** [comment]
About the options, IMO better to always use InlineContent and never data URI as it's better aligned to the spec, and in the mean time it's already been [implemented](https://github.com/NieuwlandGeo/SLDReader/issues/197) in @nieuwlandgeo/sldreader !

So last thing is to fix the tests. I rebased and could fix some tests, but still have failures that I don't understand, is it still some deprecation related things, or due to those MESA/ZINK/glx things ? (no idea what this is about!!). Either way, I don't see what's done differently in my news tests compared to existing ones. I couldn't get the failure locally (unsure if due to my sketchy docker/wsl devenv), so it's really slow to investigate, would very much appreciate some help on this :-)

```
747/993 Test #751: PyQgsSymbolLayerCreateSld ..........................***Failed    4.01 sec
export =
export PYTHONPATH=/root/QGIS/build/output/python:/root/QGIS/build/output/python/plugins:/root/QGIS/tests/src/python:
-- Running timeout -sSIGSEGV 55s /usr/bin/python3.12 /root/QGIS/tests/src/python/test_qgssymbollayer_createsld.py
MESA: error: ZINK: failed to choose pdev
glx: failed to create drisw screen
failed to load driver: zink
/root/QGIS/tests/src/python/test_qgssymbollayer_createsld.py:1776: DeprecationWarning: QgsVectorLayer.writeSld() is deprecated
  mapLayer.writeSld(root, dom, error, {})
.../root/QGIS/tests/src/python/test_qgssymbollayer_createsld.py:1564: DeprecationWarning: QgsSimpleMarkerSymbolLayer.toSld() is deprecated
  l.toSld(dom, root, dict())
```
